### ISSUE TITLE
Moving away from *Serial types

### ISSUE THREAD

**Issue by hsanjuan** [ipfs/ipfs-cluster#654]
The `*Serial` types have been used to:

a) Provide JSON representation for types in the API output
b) Have a serializable-by-default type that can also be used by Msgpack so that RPC works.

It was an early thing and I probably didn't know better. This has cascaded into converting between serializable slices and original ones etc. A mess. There's also a rather big performance impact (converting between types), and the fact that it is quite difficult to understand why things are like they are and not mess them up.

Our main troubles with serialization are:

* `cid.Cid` not serializable by default -> needs to be converted to string
* `peer.ID` are a string, but need to be humanized for JSON output.
* `time.Time` in metrics (? unsure)

If we remove all `*Serial` types, we need to make sure:

* To get efficient and working go-codec/msgpack serialization of the main types.
* To get exactly the same JSON representation

I'd propose to:

* Make `cid.Cid` implement the `https://golang.org/pkg/encoding/` interface (horrible gx bubbling) so that go-codec knows how to handle it.
* Ensure `peer.ID` `MarshalJSON` works out of the box.

There is an extra problem with Cids and JSON, as by default they encode as `{ "/" : <cid> }` which imho is the most unfriendly json encoding you could choose. The easy path is to break the API (and get aligned with how Go IPFS broke their API). The hard path is to have MarhsalJSON methods for all our types.

**Comment by hsanjuan** [comment]
Turns out protobuf supports Json https://developers.google.com/protocol-buffers/docs/proto3#json

**Comment by hsanjuan** [comment]
tl;dr: made necessary changes to go-libp2p-peer and go-cid and bubbled. #688 removes *Serial types and make the standard ones serializable by default. Protobuf is only used for permanent storage of Pins in the datastore.
